src/package-operations.ts

```diff
@@ -9,97 +9,223 @@ import {
-  }
-  return shouldBeUpdated;
+  /**
+   * Sidebar: Topological order
+   *
+   * "...a topological sort or topological ordering of a directed graph is a
+   * linear ordering of its vertices such that for every directed edge uv from
+   * vertex u to vertex v, u comes before v in the ordering."
+   * Ref: https://en.wikipedia.org/wiki/Topological_sorting
+   *
+   * In our usage, there is a directed edge from PackageA to PackageB if
+   * PackageA lists PackageB in any of the dependency fields in its
+   * `package.json`.
+   *
+   * `overallOrder` is a convention of the `dependency-graph` library, which by
+   * all appearances is the opposite of topological order.
+   * Ref: https://github.com/jriecken/dependency-graph/tree/072cdcc3eef1ee4531114e7a675e3cbe828fb33e#examples
+   */
+
+  // Traverse the monorepo's internal dependency graph in topological order.
+  return dependencyGraph
+    .overallOrder() // leaves -> root
+    .reverse() // root -> leaves (topological order)
+    .reduce((fullPackageMetadata, packageName) => {
+      const shouldUpdate = shouldUpdateMonorepoPackage(
+        packageName,
+        packagesToUpdate,
+        dependencyGraph,
+        versionSyncStrategy,
+      );
+
+      // Update the set of packages that will be updated in preparation for the
+      // next iteration.
+      if (shouldUpdate) {
+        packagesToUpdate.add(packageName);
+      }
+
+      fullPackageMetadata.set(packageName, {
+        // Every package is in the dependency graph, so we know that the key exists.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        ...intermediatePackageMetadata.get(packageName)!,
+        shouldUpdate,
+      });
+      return fullPackageMetadata;
+    }, new Map<string, PackageMetadata>());
+}
+
+/**
+ * **ATTN:** This function must be called on each monorepo package in
+ * topological order of dependency in order to return valid results.
+ *
+ * Gets whether the specified package should be updated, given the set of
+ * dependents of this package that will be updated, the monorepo dependency
+ * graph, and the active version synchronization strategy.
+ *
+ * @param packageName - The name of the package to check.
+ * @param packagesToUpdate - The (mutable) set of packages that will be updated.
+ * @param dependencyGraph - The monorepo's internal dependency graph.
+ * @param versionSyncStrategy - The version synchronization strategy.
+ * @returns Whether the package should be updated to the new version.
+ */
+export function shouldUpdateMonorepoPackage(
+  packageName: string,
+  packagesToUpdate: Set<string>,
+  dependencyGraph: DepGraph<string>,
+  versionSyncStrategy: VersionSynchronizationStrategies,
+): boolean {
+  const shouldUpdate =
+    versionSyncStrategy === VersionSynchronizationStrategies.fixed ||
+    packagesToUpdate.has(packageName) ||
+    dependencyGraph
+      .dependenciesOf(packageName)
+      .reduce<boolean>((hasChangedDependency, dependencyName) => {
```

Comment:
@rekmarks Do you mind explaining what this is doing? As I mention above, it appears that you are visiting dependents before dependencies, so if A depends on B, and B depends on C, then we'd encounter A, then B, then C. But here, we are looking up the dependencies of the package in question, which, due to the order above, we are guaranteed not to have seen yet (even taking circular dependencies into consideration). Therefore, wouldn't this always be false?

Reply:
Update: This one seems right, but there is still something off about the other comment I left that I can't square.